File: DKF/BackPropagateOutput.m
function Grads = BackPropagateOutput(StateTrue,Dynamic,States,MeasurementMinusCStates,GainMeasurementMinusCFs,MeasurementMinusCFs,TensorizedGains,MeasurementWeightMatsSym,PredictorWeightMatsSym,Grads,StateJacobians,DynJacobians,NetWeights,NetParameters)
%BACKPROPAGATEOUTPUT: Computes the gradients of the loss function with
%respects to the parameters. The loss function is:
%
%       (Penalty0/2)*||States{Layers+1}-StateTrue||^2 + 
%                          + sum_{Layer = 1,...,Layers}(Penalty1/2)*( MeasurementMinusCStates{Layer}' )*MeasurementWeightMats{Layer}*( MeasurementMinusCStates{Layer} ) + 
%                          + sum_{Layer = 1,...,Layers}(Penalty2/2)*( GainMeasurementMinusCFs{Layer}' )*PredictorWeightMats{Layer}*( GainMeasurementMinusCFs{Layer} ) +
%                          + (Penalty3/2)*||L*TensorizedGains||^2 = 
%       = Eps + sum_{Layer = 1,...,Layers} F^Layer + sum_{Layer = 1,...,Layers} G^Layer + H
%
%We can choose to backpropagate the gradients stemming from F^Layer, G^Layer or to
%truncate them at the Layer-th layer only.

%Variables
Layers = NetParameters.Layers;
C = NetParameters.C; 
LtL = NetParameters.LtL;
StateDimension = NetParameters.StateDimension;
SharedWeights = NetParameters.SharedWeights;
BackPropagation = NetParameters.BackPropagation;
Penalty0 = NetParameters.Penalty0;
Penalty1 = NetParameters.Penalty1;
Penalty2 = NetParameters.Penalty2;
Penalty3 = NetParameters.Penalty3;

GradsStateEps = cell(1);
GradsStateF = cell(1,Layers);
GradsStateG = cell(1,Layers);

%Compute the gradients with respects to the net's weights and update the 
%gradients with respects to the states along the way
for Layer = Layers:-1:1
    if strcmp(SharedWeights,'Yes')
        Indx = 1;
    end
    if strcmp(SharedWeights,'No')
        Indx = Layer;
    end

    %Construct common matrix components
    CommonMat = -NetWeights{Indx}*C;
    if Layer > 1
        CommonMatState = CommonMat*StateJacobians{Layer};
    end
    CommonMatDyn = CommonMat*DynJacobians{Layer};
    CommonMat = eye(StateDimension) + CommonMat;

    %Construct the dynamics gradient matrix
    DynMat = ( CommonMat*DynJacobians{Layer} )';

    %Construct the gradient update matrix at current layer
    if Layer > 1
        UpdateMat = ( CommonMat*StateJacobians{Layer} )';
    end

    %Gradient of H with respect to NetWeights{Indx}
    Grads{Indx} = Grads{Indx} + Penalty3*tensorprod(TensorizedGains,LtL(Layer,:),3,2);
    
    if strcmp(BackPropagation,'Complete')

        if Layer == Layers
            %Gradient of Eps with respect to state at last layer
            GradsStateEps{1} = Penalty0*( States{end} - StateTrue );
        end
        %Gradient of Eps with respect to NetWeights{Indx}
        Grads{Indx} = Grads{Indx} + GradsStateEps{1}*MeasurementMinusCFs{Layer}';
        %Gradient of Eps with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateEps{1};

        %Gradient of F^Layer with respect to state at Layer-th layer
        GradsStateF{Layer} = -Penalty1(Layer)*( C'*MeasurementWeightMatsSym{Layer}*MeasurementMinusCStates{Layer} );
        %Gradient of F^Layer with respect to NetWeights{Indx}
        Grads{Indx} = Grads{Indx} + GradsStateF{Layer}*MeasurementMinusCFs{Layer}';
        %Gradient of F^Layer with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateF{Layer};

        %Gradient of G^Layer with respect to fictitious state at Layer-th layer
        GradsStateG{Layer} = Penalty2(Layer)*PredictorWeightMatsSym{Layer}*GainMeasurementMinusCFs{Layer};
        %Gradient of G^Layer with respect to NetWeights{Indx}
        Grads{Indx} = Grads{Indx} + GradsStateG{Layer}*MeasurementMinusCFs{Layer}';
        %Gradient of G^Layer with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + CommonMatDyn'*GradsStateG{Layer};
   
        if Layer > 1
            %Update gradient of Eps with respect to state
            GradsStateEps{1} = UpdateMat*GradsStateEps{1};
            %Update gradient of F^Layer with respect to state
            GradsStateF{Layer} = UpdateMat*GradsStateF{Layer};
            %Update gradient of G^Layer with respect to state
            GradsStateG{Layer} = CommonMatState'*GradsStateG{Layer};
        end

        for PastLayer = Layer+1:Layers
            %Gradient of F^PastLayer with respect to NetWeights{Indx}
            Grads{Indx} = Grads{Indx} + GradsStateF{PastLayer}*MeasurementMinusCFs{Layer}';
            %Gradient of F^PastLayer with respect to NetWeights{end}{Dynamic}
            Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateF{PastLayer};

            %Gradient of G^PastLayer with respect to NetWeights{Indx}
            Grads{Indx} = Grads{Indx} + GradsStateG{PastLayer}*MeasurementMinusCFs{Layer}';
            %Gradient of G^PastLayer with respect to NetWeights{end}{Dynamic}
            Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateG{PastLayer};
            
            if Layer > 1
                %Update gradient of F^PastLayer with respect to state
                GradsStateF{PastLayer} = UpdateMat*GradsStateF{PastLayer};
                %Update gradient of G^PastLayer with respect to state
                GradsStateG{PastLayer} = UpdateMat*GradsStateG{PastLayer};
            end
        end
    end

    if strcmp(BackPropagation,'Truncated')
        %Adjust with gradient of Epsilon, F^Layer, G^Layer with respect to NetWeights{Indx}
        %and truncate its backpropagation.

        if Layer == Layers
            %Gradient of Eps with respect to state at last layer
            GradsStateEps{1} = Penalty0*( States{end} - StateTrue );

            %Gradient of Eps with respect to NetWeights{Indx}
            Grads{Indx} = Grads{Indx} + GradsStateEps{1}*MeasurementMinusCFs{end}';
        end
        %Gradient of Eps with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateEps{1};

        %Gradient of F^Layer with respect to state at Layer-th layer
        GradsStateF{Layer} = -Penalty1(Layer)*( C'*MeasurementWeightMatsSym{Layer}*MeasurementMinusCStates{Layer} );
        %Gradient of F^Layer with respect to NetWeights{Indx}
        Grads{Indx} = Grads{Indx} + GradsStateF{Layer}*MeasurementMinusCFs{Layer}';
        %Gradient of F^Layer with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateF{Layer};

        %Gradient of G^Layer with respect to fictitious state at Layer-th layer
        GradsStateG{Layer} = Penalty2(Layer)*PredictorWeightMatsSym{Layer}*GainMeasurementMinusCFs{Layer};
        %Gradient of G^Layer with respect to NetWeights{Indx}
        Grads{Indx} = Grads{Indx} + GradsStateG{Layer}*MeasurementMinusCFs{Layer}';
        %Gradient of G^Layer with respect to NetWeights{end}{Dynamic}
        Grads{end}{Dynamic} = Grads{end}{Dynamic} + CommonMatDyn'*GradsStateG{Layer};
   
        if Layer > 1
            %Update gradient of Eps with respect to state
            GradsStateEps{1} = UpdateMat*GradsStateEps{1};
            %Update gradient of F^Layer with respect to state
            GradsStateF{Layer} = UpdateMat*GradsStateF{Layer};
            %Update gradient of G^Layer with respect to state
            GradsStateG{Layer} = CommonMatState'*GradsStateG{Layer};
        end

        for PastLayer = Layer+1:Layers
            %Gradient of F^PastLayer with respect to NetWeights{end}{Dynamic}
            Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateF{PastLayer};

            %Gradient of G^PastLayer with respect to NetWeights{end}{Dynamic}
            Grads{end}{Dynamic} = Grads{end}{Dynamic} + DynMat*GradsStateG{PastLayer};
            
            if Layer > 1
                %Update gradient of F^PastLayer with respect to state
                GradsStateF{PastLayer} = UpdateMat*GradsStateF{PastLayer};
                %Update gradient of G^PastLayer with respect to state
                GradsStateG{PastLayer} = UpdateMat*GradsStateG{PastLayer};
            end
        end
    end
end
end

File: DKF-Python/BackPropagateOutput.py
import numpy as np

def BackPropagateOutput(StateTrue, Dynamic, States, MeasurementMinusCStates, GainMeasurementMinusCFs, MeasurementMinusCFs, TensorizedGains, MeasurementWeightMatsSym, PredictorWeightMatsSym, Grads, StateJacobians, DynJacobians, NetWeights, NetParameters):
    """
    Computes the gradients of the loss function with respect to the parameters.
    The loss function is:
    
    (Penalty0/2)*||States[Layers+1] - StateTrue||^2 + 
    sum_{Layer=1,...,Layers}(Penalty1/2)*( MeasurementMinusCStates[Layer].T )*MeasurementWeightMats[Layer]*( MeasurementMinusCStates[Layer] ) + 
    sum_{Layer=1,...,Layers}(Penalty2/2)*( GainMeasurementMinusCFs[Layer].T )*PredictorWeightMats[Layer]*( GainMeasurementMinusCFs[Layer] ) +
    (Penalty3/2)*||L*TensorizedGains||^2
    """

    Layers = NetParameters['Layers']
    C = NetParameters['C']
    LtL = NetParameters['LtL']
    StateDimension = NetParameters['StateDimension']
    SharedWeights = NetParameters['SharedWeights']
    BackPropagation = NetParameters['BackPropagation']
    Penalty0 = NetParameters['Penalty0']
    Penalty1 = NetParameters['Penalty1']
    Penalty2 = NetParameters['Penalty2']
    Penalty3 = NetParameters['Penalty3']

    GradsStateEps = [None]
    GradsStateF = [None] * Layers
    GradsStateG = [None] * Layers

    # Loop backward over the layers
    for Layer in range(Layers - 1, -1, -1):
        if SharedWeights == 'Yes':
            Indx = 0
        else:
            Indx = Layer

        # Common matrix components
        CommonMat = -NetWeights[Indx] @ C
        if Layer > 0:
            CommonMatState = CommonMat @ StateJacobians[Layer]
        CommonMatDyn = CommonMat @ DynJacobians[Layer]
        CommonMat = np.eye(StateDimension) + CommonMat

        # Dynamics gradient matrix
        DynMat = (CommonMat @ DynJacobians[Layer]).T

        # Gradient update matrix at current layer
        if Layer > 0:
            UpdateMat = (CommonMat @ StateJacobians[Layer]).T

        # Gradient of H with respect to NetWeights[Indx]
        Grads[Indx] += Penalty3 * np.tensordot(TensorizedGains, LtL[Layer,:], axes=([2], [0]))

        if BackPropagation == 'Complete':
            if Layer == Layers - 1:
                # Gradient of Eps with respect to state at last layer
                GradsStateEps[0] = Penalty0 * (States[-1] - StateTrue)
            
            # Gradient of Eps with respect to NetWeights[Indx]
            Grads[Indx] += GradsStateEps[0] @ MeasurementMinusCFs[Layer].T
            # Gradient of Eps with respect to NetWeights[-1][Dynamic]
            Grads[-1][Dynamic-1] += DynMat @ GradsStateEps[0]

            # Gradient of F^Layer
            GradsStateF[Layer] = -Penalty1[Layer] * (C.T @ MeasurementWeightMatsSym[Layer] @ MeasurementMinusCStates[Layer])
            Grads[Indx] += GradsStateF[Layer] @ MeasurementMinusCFs[Layer].T
            Grads[-1][Dynamic-1] += DynMat @ GradsStateF[Layer]

            # Gradient of G^Layer
            GradsStateG[Layer] = Penalty2[Layer] * PredictorWeightMatsSym[Layer] @ GainMeasurementMinusCFs[Layer]
            Grads[Indx] += GradsStateG[Layer] @ MeasurementMinusCFs[Layer].T
            Grads[-1][Dynamic-1] += CommonMatDyn.T @ GradsStateG[Layer]

            if Layer > 0:
                # Update gradients
                GradsStateEps[0] = UpdateMat @ GradsStateEps[0]
                GradsStateF[Layer] = UpdateMat @ GradsStateF[Layer]
                GradsStateG[Layer] = CommonMatState.T @ GradsStateG[Layer]

            # Loop over past layers for gradient accumulation
            for PastLayer in range(Layer + 1, Layers):
                Grads[Indx] += GradsStateF[PastLayer] @ MeasurementMinusCFs[Layer].T
                Grads[-1][Dynamic-1] += DynMat @ GradsStateF[PastLayer]
                Grads[Indx] += GradsStateG[PastLayer] @ MeasurementMinusCFs[Layer].T
                Grads[-1][Dynamic-1] += DynMat @ GradsStateG[PastLayer]

                if Layer > 0:
                    GradsStateF[PastLayer] = UpdateMat @ GradsStateF[PastLayer]
                    GradsStateG[PastLayer] = UpdateMat @ GradsStateG[PastLayer]

        elif BackPropagation == 'Truncated':
            if Layer == Layers - 1:
                GradsStateEps[0] = Penalty0 * (States[-1] - StateTrue)
                Grads[Indx] += GradsStateEps[0] @ MeasurementMinusCFs[-1].T

            Grads[-1][Dynamic-1] += DynMat @ GradsStateEps[0]
            GradsStateF[Layer] = -Penalty1[Layer] * (C.T @ MeasurementWeightMatsSym[Layer] @ MeasurementMinusCStates[Layer])
            Grads[Indx] += GradsStateF[Layer] @ MeasurementMinusCFs[Layer].T
            Grads[-1][Dynamic-1] += DynMat @ GradsStateF[Layer]

            GradsStateG[Layer] = Penalty2[Layer] * PredictorWeightMatsSym[Layer] @ GainMeasurementMinusCFs[Layer]
            Grads[Indx] += GradsStateG[Layer] @ MeasurementMinusCFs[Layer].T
            Grads[-1][Dynamic-1] += CommonMatDyn.T @ GradsStateG[Layer]

            if Layer > 0:
                # Update gradients with respect to the states
                GradsStateEps[0] = UpdateMat @ GradsStateEps[0]
                GradsStateF[Layer] = UpdateMat @ GradsStateF[Layer]
                GradsStateG[Layer] = CommonMatState.T @ GradsStateG[Layer]

            # Loop over past layers and update gradients
            for PastLayer in range(Layer + 1, Layers):
                Grads[-1][Dynamic-1] += DynMat @ GradsStateF[PastLayer]
                Grads[-1][Dynamic-1] += DynMat @ GradsStateG[PastLayer]

                if Layer > 0:
                    GradsStateF[PastLayer] = UpdateMat @ GradsStateF[PastLayer]
                    GradsStateG[PastLayer] = UpdateMat @ GradsStateG[PastLayer]

    return Grads
